api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java

```diff
@@ -0,0 +1,111 @@
+
+import static java.util.Collections.singletonList;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.baggage.Baggage;
+import io.opentelemetry.baggage.BaggageManager;
+import io.opentelemetry.baggage.BaggageUtils;
+import io.opentelemetry.baggage.EmptyBaggage;
+import io.opentelemetry.baggage.Entry;
+import io.opentelemetry.baggage.EntryMetadata;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import java.util.List;
+
+/**
+ * {@link TextMapPropagator} that implements the W3C specification for baggage header propagation.
+ */
+public class W3CBaggagePropagator implements TextMapPropagator {
+
+  private static final String FIELD = "baggage";
+  private static final List<String> FIELDS = singletonList(FIELD);
+  private static final W3CBaggagePropagator INSTANCE =
+      new W3CBaggagePropagator(OpenTelemetry.getBaggageManager());
+
+  private final BaggageManager baggageManager;
+
+  // visible for testing
+  W3CBaggagePropagator(BaggageManager baggageManager) {
+    this.baggageManager = baggageManager;
+  }
+
+  /**
+   * Singleton instance of the W3C Baggage Propagator. Uses the {@link BaggageManager} from the
+   * {@link OpenTelemetry} global.
+   */
+  public static W3CBaggagePropagator getInstance() {
+    return INSTANCE;
+  }
+
+  @Override
+  public List<String> fields() {
+    return FIELDS;
+  }
+
+  @Override
+  public <C> void inject(Context context, C carrier, Setter<C> setter) {
+    Baggage baggage = BaggageUtils.getBaggage(context);
+    if (baggage == null) {
+      return;
+    }
+    StringBuilder headerContent = new StringBuilder();
+    for (Entry entry : baggage.getEntries()) {
+      headerContent.append(entry.getKey()).append("=").append(entry.getValue());
+      String metadataValue = entry.getEntryMetadata().getValue();
+      if (metadataValue != null && !metadataValue.isEmpty()) {
+        headerContent.append(";").append(metadataValue);
+      }
+      headerContent.append(",");
+    }
+    if (headerContent.length() > 0) {
+      headerContent.setLength(headerContent.length() - 1);
+      setter.set(carrier, FIELD, headerContent.toString());
+    }
+  }
+
+  @Override
+  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
+    String baggageHeader = getter.get(carrier, FIELD);
+    if (baggageHeader == null) {
+      return context;
+    }
+    if (baggageHeader.isEmpty()) {
+      return BaggageUtils.withBaggage(EmptyBaggage.getInstance(), context);
+    }
+
+    Baggage.Builder baggageBuilder = baggageManager.baggageBuilder();
+    try {
+      extractEntries(baggageHeader, baggageBuilder);
+    } catch (Exception e) {
+      return BaggageUtils.withBaggage(EmptyBaggage.getInstance(), context);
+    }
+    return BaggageUtils.withBaggage(baggageBuilder.build(), context);
+  }
+
+  @SuppressWarnings("StringSplitter")
+  private static void extractEntries(String baggageHeader, Baggage.Builder baggageBuilder) {
```

Comment:
Random noise from me - not sure what the difference between baggage and tracestate is, they're both propagated K-V pairs it seems. Anyways ;)

Reply:
My feeling is: TraceState is a fixed format, semantically well-defined spec. Baggage is the wild west free-for-all for whatever users want to do.